### ISSUE TITLE
issue with LAMG and star-like graphs

### ISSUE THREAD

**Issue by kisasexypantera94** [networkit/networkit#763]
I am computing effective resistances using LAMG and I ran into a problem that if the graph is close to the form of a star, the nodes collapse into one and at the next level of the hierarchy there is a matrix of one element. Because of this the solution is zero which is not desired. Is there anything I can do about it? Tried using conjugate gradient solver but its performance is poor.

**Comment by hmeyerhenke** [comment]
How big is your star-like graph?

**Comment by kisasexypantera94** [comment]
> How big is your star-like graph?

2.5 million edges, 1.8 million nodes

**Comment by hmeyerhenke** [comment]
How long does CG take?

**Comment by hmeyerhenke** [comment]
(Admittedly our CG implementation is not top-notch, but such a size should not be super-slow.)

**Comment by avdgrinten** [comment]
If only few of the "non-central" vertices are connected, you can save a lot of work by not computing all pair-wise distances since they will be symmetric anyway. (If the graph is a "true" star (or more generally, a tree), the distances should match the usual graph distances (up to scaling)).

**Comment by kisasexypantera94** [comment]
> How long does CG take?

On regular graph with 320k nodes and 417k edges it will take more than 100 minutes even with rtol=0.3, while LAMG works in a few seconds

**Comment by hmeyerhenke** [comment]
> > How long does CG take?
> 
> On regular graph with 320k nodes and 417k edges it will take more than 100 minutes even with rtol=0.3, while LAMG works in a few seconds

Hmm, we never put a lot of thought into CG, but that is really slow... It's more of a validation thing in NetworKit. But LAMG by design is apparently not a great choice for your type of graph.

You can also try to reduce the graph by eliminating the degree-1 vertices, compute the resistances, then reinsert and readjust. Sounds like this should make things a lot faster (but needs some mild programming logic).

**Comment by hmeyerhenke** [comment]
> You can also try to reduce the graph by eliminating the degree-1 vertices, compute the resistances, then reinsert and readjust. Sounds like this should make things a lot faster (but needs some mild programming logic).

Which would be a generalization of what @avdgrinten said.

**Comment by kisasexypantera94** [comment]
yeah, it looks like this is the most reasonable thing I can do. Thank you!

**Comment by hmeyerhenke** [comment]
I think this may even work by a reduction to the 2-core...

**Comment by kisasexypantera94** [comment]
btw shouldn't LAMG warn the user when such thing happens? It also returns zero-dimensioned x, not sure if it's ok

**Comment by hmeyerhenke** [comment]
It should. What would you say would be suitable conditions to test for failure?

**Comment by kisasexypantera94** [comment]
Generate any star graph with more than 200 nodes and compute something using LAMG (commute distance for example).
This line will return zero-dimensioned vector:
https://github.com/networkit/networkit/blob/49012a90834024eaad1e1c6d70502413ba6ea9e4/include/networkit/numerics/LAMG/SolverLamg.hpp#L105

**Comment by kisasexypantera94** [comment]
Generate any star graph with more than 200 nodes and compute something using LAMG.
```c++
Graph gen_graph(size_t num_nodes) {
    Graph graph(num_nodes, true);

    for (size_t i = 1; i < num_nodes; ++i) {
        graph.addEdge(0, i, 1.0/double(5));
    }
    graph.indexEdges();

    return graph;
}

int main {
    auto graph = gen_graph(201);
    CommuteTimeDistance dist(graph);

    dist.runApproximation();
    auto d = dist.distance(0, 1);
```

This line will return zero-dimensioned vector:
https://github.com/networkit/networkit/blob/49012a90834024eaad1e1c6d70502413ba6ea9e4/include/networkit/numerics/LAMG/SolverLamg.hpp#L105

**Comment by hmeyerhenke** [comment]
Sure, a zero-dim. vector should be easy to test for (and catch). I am wondering why this happens and if there is something beforehand we can do to avoid this. But I guess this is more a general question than one posed to you necessarily, @kisasexypantera94.

**Comment by kisasexypantera94** [comment]
I'm not sure but it seems like this only happens when there is only one central node (hub), which should be pretty straightforward to check. But also need to take a closer look at the LAMG paper and implementation

**Comment by hmeyerhenke** [comment]
Probably the coarsening step leads to one node only or so. Thanks!
